File: imperative-android-hn/app/src/main/java/com/prof18/imperativateandroidhn/ui/MainViewModel.kt
package com.prof18.imperativateandroidhn.ui

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.prof18.imperativateandroidhn.data.model.AppState
import com.prof18.imperativateandroidhn.data.model.NewsState
import com.prof18.imperativateandroidhn.data.newsList
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class MainViewModel : ViewModel() {

    private val _appState = MutableLiveData<AppState>()
    val appState: LiveData<AppState>
        get() = _appState


    fun loadData() {
        viewModelScope.launch {
            _appState.value = AppState(newsState = NewsState.Loading)
            delay(2000)
            _appState.value = AppState(newsState = NewsState.Success(newsList))
        }
    }

    fun generateError() {
        viewModelScope.launch {
            delay(2000)
            _appState.value =  AppState(newsState = NewsState.Error("This is a generated error only to try an error state"))
        }
    }

}

File: imperative-ios-hn/Imperative iOs HN/ui/MainViewModel.swift
//
//  MainViewModel.swift
//  Imperative iOs HN
//
//  Created by Marco Gomiero on 14/02/2020.
//  Copyright © 2020 Marco Gomiero. All rights reserved.
//

import Foundation

class MainViewModel {
    
    func loadData() {
        let loadingAppState = AppState(newsState: Loading())
        NotificationCenter.default.post(name: NSNotification.Name(rawValue: "appState"), object: nil, userInfo: ["appState": loadingAppState])
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
            let successState = AppState(newsState: Success(news: newsList))
            NotificationCenter.default.post(name: NSNotification.Name(rawValue: "appState"), object: nil, userInfo: ["appState": successState])
        }
    }
    
    func generateError() {
        let loadingAppState = AppState(newsState: Loading())
        NotificationCenter.default.post(name: NSNotification.Name(rawValue: "appState"), object: nil, userInfo: ["appState": loadingAppState])
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
            let errorState = AppState(newsState: Error(reason: "This is a generated error only to try an error state"))
            NotificationCenter.default.post(name: NSNotification.Name(rawValue: "appState"), object: nil, userInfo: ["appState": errorState])
        }
    }
}
